### DISCUSSION TITLE
Menu/EEPROM: Factory reset support

### DISCUSSION THREAD

**Discussion by vzahradnik** [davetcc/tcMenu#296] [API & Remote]
Currently it is not possible to reset the menu to its factory state:
- There is no method to reset the menu, including menu item values
- Item is assigned a default value after bootup, but once the item value is changed, we can't read the default value anymore

There is a workaround, which involves writing a custom code to set each menu item to a default value and then saving the config. However, in this code we duplicate the information stored inside our project file. For example:
  - We set the default value in TcMenu Generator
  - We set the same default value in our custom function to factory reset the menu
![image](https://user-images.githubusercontent.com/4460766/216769769-fcfc973f-1dad-4cf3-ac1e-648defb0ff95.png)

**Comment by davetcc** [comment]
To factory reset, I normally just erase the magic key in the first two bytes and reset the device. When it comes back up the EEPROM will not load, and everything will be at the default values as if it were starting for the first time. In fact, I generally wouldn't even do the reset, just tell the user to reset in case it was not convenient.

For my case of mainly audio equipment and a few garden devices, this works well for this very occasional case.

There is also a change going into 3.1 that will allow for sized EEPROM, IE the largest location saved out is also stored, and the loader will ignore any item with a location above that. This allows for default values in new items.

**Comment by vzahradnik** [comment]
Yeah, I saw you already updated the docs but the code is not yet available. I reserved two bytes already to enable this feature.

Ok, your suggestion makes sense. Device reset is rather quick.

Thanks!

**Comment by davetcc** [comment]
I'll convert it to a discussion for now, so we can keep discussing it. How often do you think the "reset to factory" would be used? I guess that drives how feasible restarting is.

**Reply by vzahradnik** [reply]
In the production code I guess it will be rare. This is a feature more useful while developing the firmware. For now I will use your advice and just restart the device. I think it is good enough.

But I'm thinking that perhaps having a default value as an attribute could be useful. Like the `AnalogItem` has the MaxValue which I can read via a method.

Possible usecases:
1. Complete factory reset -- can be resolved with what you suggested
2. Ability to set just selected items to their default state. For example, user changed some values in the screen and decided to revert back.

Usecase 2 gives some sense from the user perspective. For me it's hard to predict how my code will be used in the future. But I can always create some issue later.

**Comment by davetcc** [comment]
I think we need to think about this a bit more, maybe a function could be provided that did the reset, then if people didn't use it, it would be optimized out by the linked. Otherwise, it would be quite a penalty on large menus.

**Reply by vzahradnik** [reply]
I agree. If we do it, it should be simple and efficient.

**Comment by vzahradnik** [comment]
So I implemented the factory reset the way you suggested and it work great for my current usecase.

I thought about possible scenarios where a library support would be helpful:
- Setting to defaults just some parts of the menu. Similarly how you can now set individual variables inside EEPROM. For example, there would be a "Restore defaults" button inside "Network" submenu and just the network settings would be restored
- Reset currently involves rebooting the microcontroller. In cases where it handles some activity (opening a door, keeping the lights open, etc.), rebooting is not the best option as the control will be interrupted for a brief moment.

Having said that, at least in my appliance I don't need these usecases at the moment. Time will tell whether my current needs change.

**Reply by davetcc** [reply]
> Having said that, at least in my appliance I don't need these usecases at the moment. Time will tell whether my current needs change.

Let's wait until a use case comes up that needs the additional support and implement it then, that way we'll have requirements to work to.
